pkg/cli/codearea.go

```diff
@@ -196,6 +200,56 @@ func (w *codeArea) handlePasteSetting(start bool) bool {
 	return true
 }
 
+// Try to expand a non-word abbreviation.
+func (w *codeArea) expandAbbr() {
+	var abbr, full string
+	// Try to expand an abbreviation, preferring the longest one
+	w.Abbreviations(func(a, f string) bool {
+		if strings.HasSuffix(w.inserts, a) && len(a) > len(abbr) {
+			abbr, full = a, f
+		}
+		return false // check the next abbr
+	})
+	if len(abbr) > 0 {
+		c := &w.State.Buffer
+		*c = CodeBuffer{
+			Content: c.Content[:c.Dot-len(abbr)] + full + c.Content[c.Dot:],
+			Dot:     c.Dot - len(abbr) + len(full),
+		}
+		w.resetInserts()
+	}
+}
+
+// Try to expand a small word abbreviation.
+func (w *codeArea) expandSmallWordAbbr() {
+	w.StateMutex.Lock()
+	defer w.StateMutex.Unlock()
+
+	ilen := len(w.inserts)
+	w.WordAbbreviations(func(a, f string) bool {
+		alen := len(a)
+		if alen > ilen { // skip abbr since it can't match the insert buffer
+			return false // check the next abbreviation
+		}
+		if ilen > alen {
+			rune, _ := utf8.DecodeLastRuneInString(w.inserts[:ilen-alen])
+			if IsSmallWordRune(rune) { // the abbreviation doesn't match a small word
```

Comment:
There are two classes of small words: a run of alphanumeric characters (e.g. `abc`) is a small word, but so is a run of non-whitespace, non-alphanumeric characters (e.g. `*/*`), so there is no single definition for what a "small word rune" is; there are two symmetric small-word categories. See [this comment](https://github.com/elves/elvish/blob/1d85b7fced66688367096767d707cc992807add6/pkg/edit/builtins.go#L462] on how this all works.

The correct way to identify whether a word abbreviation should be expanded is to detect *word boundaries*, where the category of two adjacent runes are different. You should do something like:

```
if categorizeSmallWord(r1) != categorizeSmallWord(r2) {
  // hit
}
```

Detecting word boundaries is also useful. As a theoretical example, you can now define `&` as a small-word expansion that only expands when both sides of it are alphanumerical characters. You can even mix and define `|a` as a small-word expansion that only expands when the left side of it is an alphanumerical character, but the right side of it is not.

Reply:
I like this idea. It supports my legacy "word" abbreviations imported from the `fish` shell and is more general.